### ISSUE TITLE
GPS marker on map

### ISSUE THREAD

**Issue by ljbade** [mapbox/mapbox-gl-native#756]
I want to take whatever @bleege cooks up in https://github.com/mapbox/mapbox-android-sdk/issues/593 and https://github.com/mapbox/mapbox-android-sdk/issues/592 to implement a GPS map marker in the demo app.

**Comment by ljbade** [comment]
I changed the GPS button to toggle location updates on or off. The updates are once a second and keep the map centre updated.

I used the https://github.com/mapzen/LOST that @bleege found.

Now I need to figure out how to add a marker to the map with compass direction and accuracy circle.

@incanus I guess this would be a good use of the client-side data stuff you are working on?

**Comment by incanus** [comment]
Yes, this is exactly what that is for. You can maybe fake it at screen center for tracking for now (i.e. put the marker *above* the map) but that won't help when you want to pan the map while showing the user.

**Comment by ljbade** [comment]
Yeah I was thinking about faking it @incanus. To get around the scrolling issue I may calculate screen coord from the zoom level and lat/lon

**Comment by incanus** [comment]
Yeah, for a single marker, I'll bet that will work fine for now. I'll be watching as that could be useful on iOS as a stopgap too.

**Comment by ljbade** [comment]
Will need to use these conversions:
https://github.com/mapbox/mapbox-gl-native/issues/476

**Comment by ljbade** [comment]
Related iOS ticket: https://github.com/mapbox/mapbox-gl-native/issues/147

**Comment by incanus** [comment]
Related to annotations, I'm curious @ljbade @bleege @1ec5 @kkaefer to hear your thoughts on user tracking settings and how we might push some of that down to the GL core. 

For example, on iOS, with our SDK we took the model offered by MapKit, which is to have two separate API: 

### `showsUserLocation`

A boolean that either shows the user dot or not. 

### `trackingMode`

A three-way option, the latter two options of which require or enable `showsUserLocation`: 

* `.None` — Just show (or don't) the user dot.
* `.Follow` — Center the map view on the user dot (including during movement) until the mode is changed or the map is panned manually. 
* `.FollowWithHeading` — Same as above but also keep the map rotated to the direction the user is facing (compass). 

Basically, does it make sense to do *all* of this per-platform (as we have now on Android, for example), or to have the common modes and toggles at the core level and call out to e.g. Core Location on iOS for the location functionality?

**Comment by ljbade** [comment]
@incanus: how would we feed resources into GL? e.g. different resolution marker bitmaps with different scaling depending on dpi? Android does this for free with my current approach

Is there different behaviour/visuals in respect to how Google vs Apple display GPS location?

Also I would prefer to push the GPS coords from Java to native, and not callback from native to Java to fetch them.

**Comment by incanus** [comment]
I don't mean to do the marker rendering in the core necessarily, but just the mode state tracking and calling back out to the platform edges for location API and client notifications.

**Comment by ljbade** [comment]
I have the initial version of the GPS marker (with compass support as backup) in https://github.com/mapbox/mapbox-gl-native/tree/android-conversions

I think I will merge this, and continue more improvements in another branch.

Future improvements:
- [ ] accuracy circle
- [ ] smoothly animate the marker to new positions
- [ ] update marker position during map animations
- [ ] filter compass bearings to reduce shake
- [ ] implement map position tracking and disable when the user pans the map (like Google Maps)
- [ ] add a third state that rotates the map to match the compass and disable when the user rotates the map (like Google Maps)

**Comment by jfirebaugh** [comment]
Per chat, @1ec5 to look at the iOS side of this, porting functionality from the raster iOS SDK.

**Comment by incanus** [comment]
> Per chat, @1ec5 to look at the iOS side of this, porting functionality from the raster iOS SDK.

As a first cut, we'll just do this platform-side. Later, we can push state down into the GL map layer.

**Comment by incanus** [comment]
A good summary of what's involved is anything mentioning `userTracking` in [`RMMapView.m`](https://github.com/mapbox/mapbox-ios-sdk/blob/release/MapView/Map/RMMapView.m). I don't think we should worry about an `MKUserTrackingBarButtonItem` equivalent for now.

**Comment by 1ec5** [comment]
Current status: In the 756-gps-marker branch, I adapted much of the mapbox-ios-sdk implementation but consolidated the drawing code. The delegate methods are still unimplemented, but the user dot is functional now:

![user-dot](https://cloud.githubusercontent.com/assets/1231218/6671017/9b77a4cc-cbbf-11e4-92bb-275c196e6714.png)

**Comment by ljbade** [comment]
@1ec5 so does iOS maps have the soft-lock thing where the map tracks the marker, until you touch it when it disables the lock? I see Google Maps tracks both rotation and pan locking.

I think there needs to be a bit of a deadzone so an accidental touch doesn't unlock the marker.

**Comment by 1ec5** [comment]
@ljbade, Maps (and mapbox-ios-sdk) tracks the marker until you manually pan, zoom, or rotate. If you tap the compass (to face north), the map continues to track the user dot but no longer rotates with it. A fudge factor sounds like an interesting idea, probably out of scope for beta 1.

And oh wow, I’m just noticing that we commandeered this Android ticket for iOS work. Oops.

**Comment by ljbade** [comment]
Hmm well once we are done with iOS, might as well just reassign to Android beta milestone rather than closing.

**Comment by incanus** [comment]
iOS annotations API is in as of https://github.com/mapbox/mapbox-gl-native/commit/eded1d90aa4fa607bb5bd804a9da2cdc690aa16d. Not sure yet how to best reconcile this ticket's branch and that.

**Comment by 1ec5** [comment]
Current work, fully reconciled with the iOS annotations API, is in the [756-user-dot](https://github.com/mapbox/mapbox-gl-native/compare/756-user-dot) branch, which depends on #1074.

**Comment by jfirebaugh** [comment]
Now that we've got a PR for iOS, I'm removing this from the milestone and untagging iOS, but keeping the bug open to track the Android implementation.

**Comment by bleege** [comment]
GPS Marker is now in the SDK as of #1940.  It's missing the Accuracy Circle for now, but am holding off as we need the location tech first and I want to look into redoing the User Location marker artwork to make it more [Material Design](https://www.google.com/design/spec/material-design/introduction.html) friendly.

**Comment by erf** [comment]
@incanus i've implemented tracking as you describe, and with using the material design floating action button at the bottom right to toggle the modes, with material design icons from google (https://www.google.com/design/icons/). Would you be interested in me adding this to a branch?

**Comment by bleege** [comment]
@erf This is great!  Thank you for doing this.  We'll look to add this as part of #1856 .

**Comment by erf** [comment]
Glad to help! 
I'm working on the Android version of Embark (http://letsembark.io), so i'm happy to see this project moving forward.

**Comment by ljbade** [comment]
@bleege Can we add some sort of optional heading marker arrow to the location dot like you get in Google Maps?

**Comment by ljbade** [comment]
Also we should try to get some nice compass and location marker artwork that matches Material design.

For reference in Google Maps:
![image](https://cloud.githubusercontent.com/assets/1215089/9268362/9a410888-42a1-11e5-98ec-10688bcfbce7.png)

Accuracy circle is blue with white ring and drop shadow. it has a small translucent blue ring around it. There is also a separate arrow that can be optionally enabled and shows current compass or GPS heading.

Finally there is a separate lighter translucent blue accuracy circle with a thin darker border. I think this can be drawn in the app so no need for a image here. 

The compass is smaller than what we use and partially translucent, and overall 'simpler'. When you tap on it and the map rotates to north you get a brief compass with a 'N`:
![image](https://cloud.githubusercontent.com/assets/1215089/9268383/d508d5fe-42a1-11e5-8ba0-c215794234b2.png)

This then fades out and disappears when map is fully north.

List of images:
- small white circle with background, grey border and drop shadow for compass background
- a compass red and white needle with slight shading - this will be rotated as needed in the app
- a compass red needle with black "N" - I think the N can be rendered in app or use baked image with correct Android font
- a position marker with blue circle, white border, drop shadow and darker transparent outer ring
- a small blue arrow surrounded by transparent and aligned to be able to be rotated around the position marker

If we calculate the correct virtual size, then we can work out the dimensions needed for the images at each DPI level.

@bleege Who is best to task with creating us these images?

**Comment by bleege** [comment]
@ljbade New imagery for UserLocation has [already been ticketed internally](https://github.com/mapbox/gl-internal/issues/262).  :-)

Heading / User Location Following first pass of implementation was graciously provided by @erf .  It's on hold for the time being until we can make the processing more efficient.  Please recall #2049.

**Comment by ljbade** [comment]
@bleege Ah I understand! I'm getting mixed up with these different location tickets.

**Comment by bleege** [comment]
GPS Marker has been long implemented.  Will implement Heading / User Location Following in `0.2.0`.
